src/InvariantTheory/invariant_rings.jl

```diff
@@ -0,0 +1,185 @@
+
+   modular::Bool
+
+   primary::Vector{X}
+   secondary::Vector{X}
+   irreducible_secondary::Vector{X}
+   fundamental::Vector{X}
+
+   # Cache some stuff on the Singular side
+   # (possibly removed at some point)
+   reynolds_singular::Singular.smatrix
+   molien_singular::Singular.smatrix
+   primary_singular::Singular.smatrix
+
+   function InvRing(K::S, G::U, action::Vector{V}) where {S, U, V}
+     n = degree(G)
+     R, = PolynomialRing(K, "x" => 1:n, cached = false)
+     R_sing = singular_ring(R)
+     action_singular = identity.([change_base_ring(R_sing, g) for g in action])
+     T = typeof(R)
+     X = elem_type(R)
+     W = eltype(action_singular)
+     z = new{S, T, U, V, W, X}()
+     z.field = K
+     z.poly_ring = R
+     z.group = G
+     z.action = action
+     z.action_singular = action_singular
+     z.modular = true
+     if iszero(characteristic(K))
+       z.modular = false
+     else
+       if !iszero(mod(order(G), characteristic(K)))
+         z.modular = false
+       end
+     end
+     return z
+   end
+end
+
+################################################################################
+#
+#  Field access
+#
+################################################################################
+
+coefficient_ring(I::InvRing) = I.field
+
+polynomial_ring(I::InvRing) = I.poly_ring
+
+action(I::InvRing) = I.action
+
+_action_singular(I::InvRing) = I.action_singular
+
+group(I::InvRing) = I.group
+
+ismodular(I::InvRing) = I.modular
+
+function invariant_ring(M::Vector{<: MatrixElem})
+  return invariant_ring(base_ring(M[1]), M)
+end
+
+function invariant_ring(K::Field, M::Vector{<: MatrixElem})
+  return invariant_ring(matrix_group([change_base_ring(K, g) for g in M]))
+end
+
+function invariant_ring(G::MatrixGroup)
+  n = degree(G)
+  action = mat_elem_type(typeof(G))[g.elm for g in gens(G)]
+  return InvRing(base_ring(G), G, action)
+end
+
+invariant_ring(matrices::MatrixElem{T}...) where {T} = invariant_ring(collect(matrices))
+
+function Base.show(io::IO, IR::InvRing)
+  print(io, "Invariant ring of\n")
+  print(io, group(IR), "\n")
+  print(io, "with generators\n")
+  print(io, action(IR))
+end
+
+function reynolds_molien_via_singular(IR::InvRing)
+   if !isdefined(IR, :reynolds_singular) || !isdefined(IR, :molien_singular)
+      R = polynomial_ring(IR)
+      singular_matrices = _action_singular(IR)
+      rey, mol = Singular.LibFinvar.reynolds_molien(singular_matrices...)
+      IR.reynolds_singular = rey
+      IR.molien_singular = mol
+   end
+   return IR.reynolds_singular, IR.molien_singular
+end
+
+function primary_invariants_via_singular(IR::InvRing)
+   if !isdefined(IR, :primary_singular)
+      if iszero(characteristic(coefficient_ring(IR)))
+         rey, mol = reynolds_molien_via_singular(IR)
+         P = Singular.LibFinvar.primary_char0(rey, mol)
+         R = polynomial_ring(IR)
+         p = Vector{elem_type(R)}()
+         for i = 1:ncols(P)
+            push!(p, R(P[1, i]))
```

Comment:
Based on the codecov messages, none of the code in this file is tested, not even with a trivial input. It'd be nice to get some simple tests.